optuna/integration/allennlp/_variables.py

```diff
@@ -1,22 +1,77 @@
 import os
+from typing import Optional
 
 
-_PPID = os.getppid()
-
-"""
-User might want to launch multiple studies that uses `AllenNLPExecutor`.
-Because `AllenNLPExecutor` uses environment variables for communicating
-between a parent process and a child process. A parent process creates a study,
-defines a search space, and a child process trains a AllenNLP model by
-`allennlp.commands.train.train_model`. If multiple processes use `AllenNLPExecutor`,
-the one's configuration could be loaded in the another's configuration.
-To avoid this hazard, we add ID of a parent process to each key of
-environment variables.
-"""
-_PREFIX = "{}_OPTUNA_ALLENNLP".format(_PPID)
-_MONITOR = "{}_MONITOR".format(_PREFIX)
-_PRUNER_CLASS = "{}_PRUNER_CLASS".format(_PREFIX)
-_PRUNER_KEYS = "{}_PRUNER_KEYS".format(_PREFIX)
-_STORAGE_NAME = "{}_STORAGE_NAME".format(_PREFIX)
-_STUDY_NAME = "{}_STUDY_NAME".format(_PREFIX)
-_TRIAL_ID = "{}_TRIAL_ID".format(_PREFIX)
+SPECIAL_DELIMITER = "[OPTUNA_ALLENNLP_INTEGRATION_DELIMITER]"
+OPTUNA_ALLENNLP_DISTRIBUTED_FLAG = "OPTUNA_ALLENNLP_USE_DISTRIBUTED"
+
+
+class _VariableManager:
+    """A environment variable manager for AllenNLP integration.
+
+    User might want to launch multiple studies that uses `AllenNLPExecutor`.
+    Because `AllenNLPExecutor` uses environment variables for communicating
+    between a parent process and a child process. A parent process creates a study,
+    defines a search space, and a child process trains a AllenNLP model by
+    `allennlp.commands.train.train_model`. If multiple processes use `AllenNLPExecutor`,
+    the one's configuration could be loaded in the another's configuration.
+    To avoid this hazard, we add ID of a parent process to each key of
+    environment variables.
+
+    Note that you must invoke `set_value` only in `AllenNLPExecutor`.
+    Methods in `AllenNLPPruingCallback` could be called in multiple
+    processes when enabling distributed optimization. If `set_value`
+    is invoked in the pruning callback, a consistency would break.
+    So, after initializing `AllenNLPExecutor`, `_VariableManager` provides
+    an interface to access environment variables in a read-only manner.
+
+    """
+
+    NAME_OF_KEY = {
+        "monitor": "{}_MONITOR",
+        "pruner_class": "{}_PRUNER_CLASS",
+        "pruner_keys": "{}_PRUNER_KEYS",
+        "pruner_values": "{}_PRUNER_VALUES",
+        "storage_name": "{}_STORAGE_NAME",
+        "study_name": "{}_STUDY_NAME",
+        "trial_id": "{}_TRIAL_ID",
+    }
+
+    def __init__(self, target_pid: int) -> None:
+        self.target_pid = target_pid
+
+    @property
+    def prefix(self) -> str:
+        return "{}_OPTUNA_ALLENNLP".format(self.target_pid)
+
+    def get_key(self, name: str) -> Optional[str]:
+        return self.NAME_OF_KEY.get(name)
```

Comment:
[Note] This class is only used by the developers of this integration, so I think the current design is acceptable. But, I have some comments on the implementation. We can address them in a follow-up if needed.
- The other files do not refer to this function. Can we prefix `_`?
- If we extract the key checking of `NAME_OF_KEY`, we can reduce the duplicated lines in `set_value` and `get_value`.
- Can we raise `KeyError` rather than `AssertionError`?
- If we use `@property`, the interface may be more Pythonic.

```suggestion
    def _get_key(self, name: str) -> Optional[str]:
        name_of_path = "optuna.integration.allennlp._variables._VariableManager.NAME_OF_KEY"
        if key is None:
            KeyError(f"{name} is not found in `{name_of_path}`.")
        return self.NAME_OF_KEY.get(name)
```

Reply:
Thank you for the suggestion! All comments are reasonable for me.
I revised my PR to follow your comments (except to one).

- 69cc663 makes `_VariableManager.get_key` private
- 51212d9 extracts key checking procedure to `_VariableManager._get_key`
- 40cd34b raises `KeyError` instead of `AssertionError`

I agree with you that implementing getter using property but it would generate a bunch of copied code. So for this PR, I'd keep the current (non-pythonic) interface as is. What do you think? But if you strongly think it would be better to use property, I'll revised as well. 🐱

Reply:
Looks great! Thank you for your update.